### ISSUE TITLE
Compiling on Ubuntu 18.04 x86_64 Linking CXX executable UnitTests libarmnn.so undefined reference errors

### ISSUE THREAD

**Issue by mablem8** [ARM-software/armnn#130]
I am trying to compile ARM NN for x86_64 on Ubuntu 18.04 with gcc/g++ version 7.3.0.

```bash
DIR=$HOME/sw
cd $DIR
git clone https://github.com/ARM-software/armnn.git
cd armnn/
git checkout v18.11
cd ../
mv armnn/ armnn-x86_64/
cd armnn-x86_64/
mkdir install
mkdir build
cd build/
cmake -D ARMCOMPUTENEON=1 -D ARMCOMPUTECL=1 -D BUILD_CAFFE_PARSER=1 \
 -D PROTOBUF_ROOT=$DIR/protobuf-x86_64/install \
 -D PROTOBUF_LIBRARY_DEBUG=$DIR/protobuf-x86_64/install/lib/libprotobuf.so \
 -D PROTOBUF_LIBRARY_RELEASE=$DIR/protobuf-x86_64/install/lib/libprotobuf.so \
 -D BOOST_ROOT=$DIR/boost-x86_64/install \
 -D CAFFE_GENERATED_SOURCES=$DIR/caffe-x86_64/build/src \
 -D ARMCOMPUTE_ROOT=$DIR/ComputeLibrary-x86_64 \
 -D ARMCOMPUTE_BUILD_DIR=$DIR/ComputeLibrary-x86_64/build \
 -D CMAKE_INSTALL_PREFIX=$DIR/armnn-x86_64/install \
 -D CMAKE_C_FLAGS="${CMAKE_C_FLAGS} -fPIC" \
 -D CMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS} -fPIC" ../
```

CMake output:
<details>
  <summary>Expand</summary>

```
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
--   system
--   filesystem
--   log
--   program_options
--   date_time
--   log_setup
--   thread
--   regex
--   chrono
--   atomic
-- Looking for valgrind/memcheck.h
-- Looking for valgrind/memcheck.h - not found
-- Tensorflow parser support is disabled
-- Tensorflow Lite parser support is disabled
Including backend common library into the build: $DIR/armnn-x86_64/src/backends/aclCommon/common.cmake
Including backend common library into the build: $DIR/armnn-x86_64/src/backends/backendsCommon/common.cmake
Including backend into the build: $DIR/armnn-x86_64/src/backends/cl/backend.cmake
Including backend into the build: $DIR/armnn-x86_64/src/backends/neon/backend.cmake
Including backend into the build: $DIR/armnn-x86_64/src/backends/reference/backend.cmake
Adding object library dependency to armnn: armnnAclCommon
Adding object library dependency to armnn: armnnBackendsCommon
Adding object library dependency to armnn: armnnClBackend
Adding object library dependency to armnn: armnnClBackendWorkloads
Adding object library dependency to armnn: armnnNeonBackend
Adding object library dependency to armnn: armnnNeonBackendWorkloads
Adding object library dependency to armnn: armnnRefBackend
Adding object library dependency to armnn: armnnRefBackendWorkloads
Adding object library dependency to UnitTests: armnnAclCommonUnitTests
Adding object library dependency to UnitTests: armnnBackendsCommonUnitTests
Adding object library dependency to UnitTests: armnnClBackendUnitTests
Adding object library dependency to UnitTests: armnnNeonBackendUnitTests
Adding object library dependency to UnitTests: armnnRefBackendUnitTests
-- Configuring done
-- Generating done
-- Build files have been written to: $DIR/armnn-x86_64/build
```

</details>

Continuing:

```bash
make -j$(nproc)
```

Compilation completes to 100%, then fails at `[100%] Linking CXX executable UnitTests`:
<details>
  <summary>Expand</summary>

```
[100%] Linking CXX executable UnitTests
libarmnn.so: undefined reference to `vtable for arm_compute::NEGEMMLowpMatrixMultiplyCore'
libarmnn.so: undefined reference to `arm_compute::NEPixelWiseMultiplication::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, float, arm_compute::ConvertPolicy, arm_compute::RoundingPolicy)'
libarmnn.so: undefined reference to `vtable for arm_compute::NEWidthConcatenateLayer'
libarmnn.so: undefined reference to `arm_compute::NEBatchNormalizationLayer::configure(arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, float, arm_compute::ActivationLayerInfo)'
libarmnn.so: undefined reference to `arm_compute::NEReshapeLayer::configure(arm_compute::ITensor const*, arm_compute::ITensor*)'
libarmnn.so: undefined reference to `arm_compute::NEPoolingLayer::NEPoolingLayer()'
libarmnn.so: undefined reference to `vtable for arm_compute::NENormalizationLayer'
libarmnn.so: undefined reference to `arm_compute::NENormalizationLayer::run()'
libarmnn.so: undefined reference to `typeinfo for arm_compute::INESimpleFunction'
libarmnn.so: undefined reference to `arm_compute::NEBatchNormalizationLayer::NEBatchNormalizationLayer()'
libarmnn.so: undefined reference to `vtable for arm_compute::NEGEMM'
libarmnn.so: undefined reference to `arm_compute::NEActivationLayer::configure(arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ActivationLayerInfo)'
libarmnn.so: undefined reference to `arm_compute::NEPoolingLayer::run()'
libarmnn.so: undefined reference to `arm_compute::INESimpleFunction::INESimpleFunction()'
libarmnn.so: undefined reference to `arm_compute::NEPermute::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::Strides const&)'
libarmnn.so: undefined reference to `vtable for arm_compute::INESimpleFunction'
libarmnn.so: undefined reference to `arm_compute::NEConcatenateLayer::run()'
libarmnn.so: undefined reference to `vtable for arm_compute::NELSTMLayer'
libarmnn.so: undefined reference to `arm_compute::NEFloor::configure(arm_compute::ITensor const*, arm_compute::ITensor*)'
libarmnn.so: undefined reference to `arm_compute::NEDepthwiseConvolutionLayer3x3::NEDepthwiseConvolutionLayer3x3()'
libarmnn.so: undefined reference to `arm_compute::NEPermute::configure(arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::Strides const&)'
libarmnn.so: undefined reference to `arm_compute::NEPoolingLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::PoolingLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NENormalizationLayer::configure(arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::NormalizationLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NEConcatenateLayer::NEConcatenateLayer()'
libarmnn.so: undefined reference to `arm_compute::NEFullyConnectedLayer::configure(arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::FullyConnectedLayerInfo)'
libarmnn.so: undefined reference to `arm_compute::NENormalizationLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::NormalizationLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NEBatchNormalizationLayer::run()'
libarmnn.so: undefined reference to `arm_compute::NELSTMLayer::configure(arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::LSTMParams<arm_compute::ITensor> const&, arm_compute::ActivationLayerInfo const&, float, float)'
libarmnn.so: undefined reference to `arm_compute::NEArithmeticAddition::configure(arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ConvertPolicy)'
libarmnn.so: undefined reference to `arm_compute::NEConcatenateLayer::configure(std::vector<arm_compute::ITensor*, std::allocator<arm_compute::ITensor*> > const&, arm_compute::ITensor*, arm_compute::DataLayoutDimension)'
libarmnn.so: undefined reference to `arm_compute::INESimpleFunction::run()'
libarmnn.so: undefined reference to `vtable for arm_compute::NEFullyConnectedLayer'
libarmnn.so: undefined reference to `arm_compute::NENormalizationLayer::NENormalizationLayer(std::shared_ptr<arm_compute::IMemoryManager>)'
libarmnn.so: undefined reference to `arm_compute::NEDepthwiseConvolutionLayer3x3::configure(arm_compute::ITensor*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::PadStrideInfo const&, unsigned int, arm_compute::ActivationLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NEConvolutionLayer::NEConvolutionLayer(std::shared_ptr<arm_compute::IMemoryManager>)'
libarmnn.so: undefined reference to `arm_compute::NELSTMLayer::run()'
libarmnn.so: undefined reference to `arm_compute::NELSTMLayer::NELSTMLayer(std::shared_ptr<arm_compute::IMemoryManager>)'
libarmnn.so: undefined reference to `arm_compute::NEConvolutionLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::PadStrideInfo const&, arm_compute::WeightsInfo const&, arm_compute::Size2D const&, arm_compute::ActivationLayerInfo const&, bool, unsigned int)'
libarmnn.so: undefined reference to `arm_compute::NEFullyConnectedLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::FullyConnectedLayerInfo)'
libarmnn.so: undefined reference to `arm_compute::NESoftmaxLayer::run()'
libarmnn.so: undefined reference to `arm_compute::NESoftmaxLayer::configure(arm_compute::ITensor*, arm_compute::ITensor*, float, unsigned long)'
libarmnn.so: undefined reference to `arm_compute::NEBatchNormalizationLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, float, arm_compute::ActivationLayerInfo)'
libarmnn.so: undefined reference to `arm_compute::NEDepthwiseConvolutionLayer::NEDepthwiseConvolutionLayer()'
libarmnn.so: undefined reference to `arm_compute::NEDepthwiseConvolutionLayer::configure(arm_compute::ITensor*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::PadStrideInfo const&, unsigned int, arm_compute::ActivationLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NEArithmeticSubtraction::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ConvertPolicy)'
libarmnn.so: undefined reference to `arm_compute::NEArithmeticAddition::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ConvertPolicy)'
libarmnn.so: undefined reference to `arm_compute::NEFullyConnectedLayer::run()'
libarmnn.so: undefined reference to `arm_compute::NEFullyConnectedLayer::prepare()'
libarmnn.so: undefined reference to `arm_compute::NEConcatenateLayer::validate(std::vector<arm_compute::ITensorInfo*, std::allocator<arm_compute::ITensorInfo*> > const&, arm_compute::ITensorInfo const*, arm_compute::DataLayoutDimension)'
libarmnn.so: undefined reference to `vtable for arm_compute::NESoftmaxLayer'
libarmnn.so: undefined reference to `arm_compute::NESoftmaxLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, float, unsigned long)'
libarmnn.so: undefined reference to `arm_compute::NEFullyConnectedLayer::NEFullyConnectedLayer(std::shared_ptr<arm_compute::IMemoryManager>)'
libarmnn.so: undefined reference to `arm_compute::NEArithmeticSubtraction::configure(arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ConvertPolicy)'
libarmnn.so: undefined reference to `arm_compute::NEConvolutionLayer::configure(arm_compute::ITensor*, arm_compute::ITensor const*, arm_compute::ITensor const*, arm_compute::ITensor*, arm_compute::PadStrideInfo const&, arm_compute::WeightsInfo const&, arm_compute::Size2D const&, arm_compute::ActivationLayerInfo const&, bool, unsigned int)'
libarmnn.so: undefined reference to `vtable for arm_compute::NEGEMMAssemblyDispatch'
libarmnn.so: undefined reference to `arm_compute::NEDepthwiseConvolutionLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::PadStrideInfo const&, unsigned int, arm_compute::ActivationLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NEActivationLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ActivationLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NEPoolingLayer::configure(arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::PoolingLayerInfo const&)'
libarmnn.so: undefined reference to `arm_compute::NELSTMLayer::validate(arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::ITensorInfo const*, arm_compute::LSTMParams<arm_compute::ITensorInfo> const&, arm_compute::ActivationLayerInfo const&, float, float)'
libarmnn.so: undefined reference to `arm_compute::NESoftmaxLayer::NESoftmaxLayer(std::shared_ptr<arm_compute::IMemoryManager>)'
libarmnn.so: undefined reference to `vtable for arm_compute::NEConcatenateLayer'
libarmnn.so: undefined reference to `arm_compute::NEPixelWiseMultiplication::configure(arm_compute::ITensor*, arm_compute::ITensor*, arm_compute::ITensor*, float, arm_compute::ConvertPolicy, arm_compute::RoundingPolicy)'
collect2: error: ld returned 1 exit status
CMakeFiles/UnitTests.dir/build.make:853: recipe for target 'UnitTests' failed
make[2]: *** [UnitTests] Error 1
CMakeFiles/Makefile2:226: recipe for target 'CMakeFiles/UnitTests.dir/all' failed
make[1]: *** [CMakeFiles/UnitTests.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

</details>

Any suggestions?

**Comment by mablem8** [comment]
Solution:

```bash
cmake -D ARMCOMPUTECL=1 -D BUILD_CAFFE_PARSER=1 \
 -D PROTOBUF_ROOT=$DIR/protobuf-x86_64/install \
 -D PROTOBUF_LIBRARY_DEBUG=$DIR/protobuf-x86_64/install/lib/libprotobuf.so \
 -D PROTOBUF_LIBRARY_RELEASE=$DIR/protobuf-x86_64/install/lib/libprotobuf.so \
 -D BOOST_ROOT=$DIR/boost-x86_64/install \
 -D CAFFE_GENERATED_SOURCES=$DIR/caffe-x86_64/build/src \
 -D ARMCOMPUTE_ROOT=$DIR/ComputeLibrary-x86_64 \
 -D ARMCOMPUTE_BUILD_DIR=$DIR/ComputeLibrary-x86_64/build \
 -D CMAKE_INSTALL_PREFIX=$DIR/armnn-x86_64/install \
 -D CMAKE_C_FLAGS="${CMAKE_C_FLAGS} -std=c11 -fPIC" \
 -D CMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS} -std=c++11 -fPIC" ../
```

Simple mistake, `ARMCOMPUTENEON=1` should not be used for x86_64.

**Comment by MatthewARM** [comment]
Thanks @mablem8, I'm just curious about what you're doing with ArmNN on x86_64 -  are you going to try the OpenCL backend? Or are you just testing out the front end?

**Comment by MatthewARM** [comment]
Hi @mablem8 thanks for this. I agree it would be nice to use the OpenCL backend of ArmNN on x86_64 as a development platform, but when I've tried it before I've found that there's just so much variance across OpenCL implementations that it would be quite hard to maintain compatibility. 

It's possible that if you ask the Compute Library folks about this that they might have some suggestions.

These days I just use the ArmNN reference backend for development. It's slow compared to the optimised backends but it's easy to compile and run anywhere.

Good luck with your development!
Matthew
